addons/stock/models/stock_location.py

```diff
@@ -95,6 +97,11 @@ def _compute_next_inventory_date(self):
             else:
                 location.next_inventory_date = False
 
+    @api.depends('location_id.warehouse_id', 'warehouse_view_ids')
+    def _compute_warehouse_id(self):
+        for loc in self:
+            loc.warehouse_id = self.env['stock.warehouse'].search([('view_location_id', 'parent_of', loc.id)], limit=1)
```

Comment:
Don't know if this is a real performance issue but what about something like this : 
```python
warehouses = self.env['stock.warehouse'].search([('view_location_id', 'parent_of', self.ids)])
view_by_wh = {wh.view_location_id: wh.id for wh in warehouses}
self.warehouse_id = False
for loc in self:
    for key, value in view_by_wh.items():
        if '/' + str(key) + '/' in loc.parent_path: # I guess there is better way to do this
           loc.warehouse_id = value
           continue
```
I think doing one `search` at the time could become slow with lots of locations